File: lib/sidereon/reliability.ex
defmodule Sidereon.Reliability do
  @moduledoc """
  Classical Baarda/Teunissen reliability diagnostics.

  This module exposes pre-data reliability design and ARAIM reliability reports
  from the Rust core. Inputs are range design rows or ARAIM geometry plus an
  integrity support model. Outputs include redundancy numbers, minimal
  detectable bias, external effects, and aggregate degrees of freedom.
  """

  alias __MODULE__.{
    ObservationReliability,
    RangeReliabilityRow,
    ReliabilityOptions,
    ReliabilityReport,
    ReliabilitySummary
  }

  alias Sidereon.GNSS.ARAIM
  alias Sidereon.NIF

  defmodule ReliabilityOptions do
    @moduledoc """
    Options for classical reliability calculations.

    `:alpha` is the two-sided false-alarm probability for the w-test. `:power`
    is the target detection power. `:lambda0_override` may supply a precomputed
    noncentrality parameter. `:min_redundancy` is the reporting floor below
    which an observation is marked uncheckable.
    """

    @enforce_keys [:alpha, :power, :lambda0_override, :min_redundancy]
    defstruct [:alpha, :power, :lambda0_override, :min_redundancy]

    @type t :: %__MODULE__{
            alpha: float(),
            power: float(),
            lambda0_override: float() | nil,
            min_redundancy: float()
          }

    @doc """
    Return the core default reliability options.
    """
    @spec default() :: t()
    def default do
      {alpha, power, lambda0_override, min_redundancy} = NIF.reliability_default_options()

      %__MODULE__{
        alpha: alpha,
        power: power,
        lambda0_override: lambda0_override,
        min_redundancy: min_redundancy
      }
    end

    @doc """
    Build reliability options from a keyword list.
    """
    @spec new(keyword()) :: t()
    def new(opts \\ []) when is_list(opts) do
      default = default()

      %__MODULE__{
        alpha: Keyword.get(opts, :alpha, default.alpha) / 1.0,
        power: Keyword.get(opts, :power, default.power) / 1.0,
        lambda0_override: optional_float(Keyword.get(opts, :lambda0_override, default.lambda0_override)),
        min_redundancy: Keyword.get(opts, :min_redundancy, default.min_redundancy) / 1.0
      }
    end

    @doc false
    @spec to_nif_tuple(t()) :: tuple()
    def to_nif_tuple(%__MODULE__{} = options) do
      {
        options.alpha,
        options.power,
        options.lambda0_override,
        options.min_redundancy
      }
    end

    defp optional_float(nil), do: nil
    defp optional_float(value), do: value / 1.0
  end

  defmodule RangeReliabilityRow do
    @moduledoc """
    One range row for pre-data reliability design.

    `:design_row` is the linearized measurement row for the estimated state.
    `:sigma_m` is the supplied one-sigma range model in meters.
    """

    @enforce_keys [:id, :design_row, :sigma_m]
    defstruct [:id, :design_row, :sigma_m]

    @type t :: %__MODULE__{
            id: String.t(),
            design_row: [float()],
            sigma_m: float()
          }

    @doc """
    Build a range reliability row.
    """
    @spec new(String.t(), [number()], number()) :: t()
    def new(id, design_row, sigma_m) when is_binary(id) and is_list(design_row) do
      %__MODULE__{
        id: id,
        design_row: Enum.map(design_row, &(&1 / 1.0)),
        sigma_m: sigma_m / 1.0
      }
    end

    @doc false
    @spec to_nif_tuple(t()) :: tuple()
    def to_nif_tuple(%__MODULE__{} = row) do
      {row.id, row.design_row, row.sigma_m}
    end
  end

  defmodule ObservationReliability do
    @moduledoc """
    Reliability diagnostics for one observation.

    When `:uncheckable` is true, `:mdb_m`, `:external_enu_m`, and
    `:bias_to_noise` are `nil`.
    """

    @enforce_keys [:id, :redundancy, :mdb_m, :external_enu_m, :bias_to_noise, :uncheckable]
    defstruct [:id, :redundancy, :mdb_m, :external_enu_m, :bias_to_noise, :uncheckable]

    @type t :: %__MODULE__{
            id: String.t(),
            redundancy: float(),
            mdb_m: float() | nil,
            external_enu_m: {float(), float(), float()} | nil,
            bias_to_noise: float() | nil,
            uncheckable: boolean()
          }
  end

  defmodule ReliabilitySummary do
    @moduledoc """
    Aggregate reliability diagnostics for a design.
    """

    @enforce_keys [
      :n_obs,
      :n_params,
      :dof,
      :sum_redundancy,
      :lambda0,
      :max_mdb_m,
      :min_redundancy,
      :n_uncheckable
    ]
    defstruct [
      :n_obs,
      :n_params,
      :dof,
      :sum_redundancy,
      :lambda0,
      :max_mdb_m,
      :min_redundancy,
      :n_uncheckable
    ]

    @type t :: %__MODULE__{
            n_obs: non_neg_integer(),
            n_params: non_neg_integer(),
            dof: non_neg_integer(),
            sum_redundancy: float(),
            lambda0: float(),
            max_mdb_m: {String.t(), float()} | nil,
            min_redundancy: {String.t(), float()},
            n_uncheckable: non_neg_integer()
          }
  end

  defmodule ReliabilityReport do
    @moduledoc """
    Full reliability report.
    """

    @enforce_keys [:per_observation, :summary]
    defstruct [:per_observation, :summary]

    @type t :: %__MODULE__{
            per_observation: [ObservationReliability.t()],
            summary: ReliabilitySummary.t()
          }
  end

  @type reliability_error ::
          :invalid_probability
          | :invalid_weight
          | :invalid_reliability_parameter
          | :invalid_design
          | :singular_geometry
          | :insufficient_geometry
          | :invalid_ism
          | :invalid_allocation
          | :numerical_failure
          | term()

  @doc """
  Compute Baarda's W-test noncentrality from false-alarm probability and power.

  Returns `{:ok, %{delta0: delta0, lambda0: lambda0}}`.
  """
  @spec wtest_noncentrality(number(), number()) ::
          {:ok, %{delta0: float(), lambda0: float()}} | {:error, reliability_error()}
  def wtest_noncentrality(alpha, power) do
    case NIF.reliability_wtest_noncentrality(alpha / 1.0, power / 1.0) do
      {:ok, result} -> {:ok, result}
      {:error, _reason} = err -> err
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  @doc """
  Compute reliability from caller-supplied range design rows.
  """
  @spec reliability_design([RangeReliabilityRow.t()], ReliabilityOptions.t() | keyword()) ::
          {:ok, ReliabilityReport.t()} | {:error, reliability_error()}
  def reliability_design(rows, options \\ ReliabilityOptions.default()) when is_list(rows) do
    case NIF.reliability_design(Enum.map(rows, &RangeReliabilityRow.to_nif_tuple/1), options_tuple(options)) do
      {:ok, report} -> {:ok, report(report)}
      {:error, _reason} = err -> err
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  @doc """
  Compute reliability for ARAIM geometry using the supplied ISM range model.
  """
  @spec reliability_araim(ARAIM.Geometry.t(), ARAIM.Ism.t(), ReliabilityOptions.t() | keyword()) ::
          {:ok, ReliabilityReport.t()} | {:error, reliability_error()}
  def reliability_araim(%ARAIM.Geometry{} = geometry, %ARAIM.Ism{} = ism, options \\ ReliabilityOptions.default()) do
    with {:ok, {rows, receiver, clock_systems}} <- ARAIM.Geometry.to_nif_terms(geometry),
         {:ok, {constellations, satellites}} <- ARAIM.Ism.to_nif_terms(ism) do
      case NIF.reliability_araim(
             rows,
             receiver,
             clock_systems,
             constellations,
             satellites,
             options_tuple(options)
           ) do
        {:ok, result} -> {:ok, report(result)}
        {:error, _reason} = err -> err
      end
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  defp options_tuple(%ReliabilityOptions{} = options), do: ReliabilityOptions.to_nif_tuple(options)

  defp options_tuple(options) when is_list(options),
    do: options |> ReliabilityOptions.new() |> ReliabilityOptions.to_nif_tuple()

  defp report(fields) do
    %ReliabilityReport{
      per_observation: Enum.map(fields.per_observation, &observation/1),
      summary: summary(fields.summary)
    }
  end

  defp observation(fields) do
    %ObservationReliability{
      id: fields.id,
      redundancy: fields.redundancy,
      mdb_m: fields.mdb_m,
      external_enu_m: fields.external_enu_m,
      bias_to_noise: fields.bias_to_noise,
      uncheckable: fields.uncheckable
    }
  end

  defp summary(fields) do
    %ReliabilitySummary{
      n_obs: fields.n_obs,
      n_params: fields.n_params,
      dof: fields.dof,
      sum_redundancy: fields.sum_redundancy,
      lambda0: fields.lambda0,
      max_mdb_m: fields.max_mdb_m,
      min_redundancy: fields.min_redundancy,
      n_uncheckable: fields.n_uncheckable
    }
  end
end

File: native/sidereon_nif/src/reliability.rs
//! Rustler boundary for classical reliability diagnostics.
//!
//! The Baarda noncentrality, MDB, redundancy, and external reliability
//! calculations live in `sidereon_core::quality`. This module only converts
//! Elixir terms to core inputs and encodes the core report back.

use rustler::{Encoder, Env, NifResult, Term};
use sidereon_core::araim::{AraimError, AraimGeometry, Ism};
use sidereon_core::quality::{
    reliability_araim as core_reliability_araim, reliability_design as core_reliability_design,
    wtest_noncentrality as core_wtest_noncentrality, ObservationReliability, QualityError,
    RangeReliabilityRow, ReliabilityOptions, ReliabilityReport, ReliabilitySummary,
};

type DesignRowTerm = (String, Vec<f64>, f64);
type OptionsTerm = (f64, f64, Option<f64>, f64);
type Vec3 = (f64, f64, f64);

mod atoms {
    rustler::atoms! {
        ok,
        error,
        invalid_elevation,
        missing_cn0,
        invalid_parameter,
        invalid_probability,
        invalid_system_count,
        invalid_dof,
        invalid_weight,
        invalid_reliability_parameter,
        invalid_residuals,
        invalid_design,
        singular_geometry,
        insufficient_geometry,
        unmonitorable_fault_mass,
        numerical_failure,
        invalid_ism,
        invalid_allocation
    }
}

#[derive(Debug, Clone, rustler::NifMap)]
struct WTestTerm {
    delta0: f64,
    lambda0: f64,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct ObservationReliabilityTerm {
    id: String,
    redundancy: f64,
    mdb_m: Option<f64>,
    external_enu_m: Option<Vec3>,
    bias_to_noise: Option<f64>,
    uncheckable: bool,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct ReliabilitySummaryTerm {
    n_obs: i64,
    n_params: i64,
    dof: i64,
    sum_redundancy: f64,
    lambda0: f64,
    max_mdb_m: Option<(String, f64)>,
    min_redundancy: (String, f64),
    n_uncheckable: i64,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct ReliabilityReportTerm {
    per_observation: Vec<ObservationReliabilityTerm>,
    summary: ReliabilitySummaryTerm,
}

fn decode_design_row((id, design_row, sigma_m): DesignRowTerm) -> RangeReliabilityRow {
    RangeReliabilityRow {
        id,
        design_row,
        sigma_m,
    }
}

fn decode_options(
    (alpha, power, lambda0_override, min_redundancy): OptionsTerm,
) -> ReliabilityOptions {
    ReliabilityOptions {
        alpha,
        beta: 1.0 - power,
        lambda0_override,
        min_redundancy,
    }
}

fn quality_error_atom(error: QualityError) -> rustler::Atom {
    match error {
        QualityError::InvalidElevation => atoms::invalid_elevation(),
        QualityError::MissingCn0 => atoms::missing_cn0(),
        QualityError::InvalidParameter => atoms::invalid_parameter(),
        QualityError::InvalidProbability => atoms::invalid_probability(),
        QualityError::InvalidSystemCount => atoms::invalid_system_count(),
        QualityError::InvalidDof => atoms::invalid_dof(),
        QualityError::InvalidWeight => atoms::invalid_weight(),
        QualityError::InvalidReliabilityParameter => atoms::invalid_reliability_parameter(),
        QualityError::InvalidResiduals => atoms::invalid_residuals(),
        QualityError::InvalidDesign => atoms::invalid_design(),
        QualityError::SingularGeometry => atoms::singular_geometry(),
    }
}

fn araim_error_atom(error: AraimError) -> rustler::Atom {
    match error {
        AraimError::InsufficientGeometry => atoms::insufficient_geometry(),
        AraimError::UnmonitorableFaultMass => atoms::unmonitorable_fault_mass(),
        AraimError::NumericalFailure => atoms::numerical_failure(),
        AraimError::InvalidIsm => atoms::invalid_ism(),
        AraimError::InvalidAllocation => atoms::invalid_allocation(),
    }
}

fn vec3(values: [f64; 3]) -> Vec3 {
    (values[0], values[1], values[2])
}

fn observation_term(value: ObservationReliability) -> ObservationReliabilityTerm {
    ObservationReliabilityTerm {
        id: value.id,
        redundancy: value.redundancy,
        mdb_m: value.mdb_m,
        external_enu_m: value.external_enu_m.map(vec3),
        bias_to_noise: value.bias_to_noise,
        uncheckable: value.uncheckable,
    }
}

fn usize_term(value: usize) -> i64 {
    value as i64
}

fn summary_term(value: ReliabilitySummary) -> ReliabilitySummaryTerm {
    ReliabilitySummaryTerm {
        n_obs: usize_term(value.n_obs),
        n_params: usize_term(value.n_params),
        dof: usize_term(value.dof),
        sum_redundancy: value.sum_redundancy,
        lambda0: value.lambda0,
        max_mdb_m: value.max_mdb_m,
        min_redundancy: value.min_redundancy,
        n_uncheckable: usize_term(value.n_uncheckable),
    }
}

fn report_term(value: ReliabilityReport) -> ReliabilityReportTerm {
    ReliabilityReportTerm {
        per_observation: value
            .per_observation
            .into_iter()
            .map(observation_term)
            .collect(),
        summary: summary_term(value.summary),
    }
}

fn geometry_from_terms(
    rows: Vec<crate::araim::RowTerm>,
    receiver: crate::araim::ReceiverTerm,
    clock_systems: Vec<String>,
) -> NifResult<AraimGeometry> {
    Ok(AraimGeometry {
        rows: rows
            .into_iter()
            .map(crate::araim::decode_row)
            .collect::<NifResult<_>>()?,
        receiver: crate::araim::decode_receiver(receiver)?,
        clock_systems: clock_systems
            .iter()
            .map(|system| crate::araim::system_from_term(system))
            .collect::<NifResult<_>>()?,
    })
}

fn ism_from_terms(
    constellations: Vec<crate::araim::ConstellationTerm>,
    satellites: Vec<crate::araim::SatelliteTerm>,
) -> NifResult<Ism> {
    Ok(Ism::new(
        constellations
            .into_iter()
            .map(crate::araim::decode_constellation)
            .collect::<NifResult<_>>()?,
        satellites
            .into_iter()
            .map(crate::araim::decode_satellite)
            .collect::<NifResult<_>>()?,
    ))
}

/// Return the core default reliability options as `{alpha, power, lambda0, min_redundancy}`.
#[rustler::nif]
fn reliability_default_options() -> OptionsTerm {
    let options = ReliabilityOptions::default();
    (
        options.alpha,
        1.0 - options.beta,
        options.lambda0_override,
        options.min_redundancy,
    )
}

/// Compute Baarda's W-test noncentrality from false-alarm probability and detection power.
#[rustler::nif]
fn reliability_wtest_noncentrality<'a>(env: Env<'a>, alpha: f64, power: f64) -> Term<'a> {
    match core_wtest_noncentrality(alpha, 1.0 - power) {
        Ok(lambda0) => (
            atoms::ok(),
            WTestTerm {
                delta0: lambda0.sqrt(),
                lambda0,
            },
        )
            .encode(env),
        Err(error) => (atoms::error(), quality_error_atom(error)).encode(env),
    }
}

/// Compute reliability from a caller-supplied weighted design.
#[rustler::nif(schedule = "DirtyCpu")]
fn reliability_design<'a>(
    env: Env<'a>,
    rows: Vec<DesignRowTerm>,
    options: OptionsTerm,
) -> Term<'a> {
    let rows = rows.into_iter().map(decode_design_row).collect::<Vec<_>>();
    match core_reliability_design(&rows, &decode_options(options)) {
        Ok(report) => (atoms::ok(), report_term(report)).encode(env),
        Err(error) => (atoms::error(), quality_error_atom(error)).encode(env),
    }
}

/// Compute reliability from ARAIM geometry and ISM inputs.
#[rustler::nif(schedule = "DirtyCpu")]
fn reliability_araim<'a>(
    env: Env<'a>,
    rows: Vec<crate::araim::RowTerm>,
    receiver: crate::araim::ReceiverTerm,
    clock_systems: Vec<String>,
    constellations: Vec<crate::araim::ConstellationTerm>,
    satellites: Vec<crate::araim::SatelliteTerm>,
    options: OptionsTerm,
) -> NifResult<Term<'a>> {
    let geometry = geometry_from_terms(rows, receiver, clock_systems)?;
    let ism = ism_from_terms(constellations, satellites)?;
    Ok(
        match core_reliability_araim(&geometry, &ism, &decode_options(options)) {
            Ok(report) => (atoms::ok(), report_term(report)).encode(env),
            Err(error) => (atoms::error(), araim_error_atom(error)).encode(env),
        },
    )
}
